src/test/java/book/store/onlinebookstore/shoppingcart/ShoppingCartServiceTest.java

```diff
@@ -0,0 +1,257 @@
+        Mockito.when(shoppingCartRepository.findById(user.getId())).thenReturn(Optional.empty());
+        Mockito.when(userRepository.getReferenceById(user.getId())).thenReturn(user);
+        Mockito.when(userRepository.findUserByEmail(user.getEmail()))
+                .thenReturn(Optional.of(user));
+        Mockito.when(shoppingCartRepository.save(shoppingCart)).thenReturn(shoppingCart);
+        Mockito.when(shoppingCartMapper.toShoppingCartDto(shoppingCart)).thenReturn(expected);
+        Mockito.when(authentication.getName()).thenReturn(user.getEmail());
+
+        //when
+        ShoppingCartDto actual = shoppingCartService.addCartItem(requestDto, authentication);
+
+        //then
+        Assertions.assertEquals(expected, actual);
+        Mockito.verify(shoppingCartRepository, Mockito.times(1))
+                .save(shoppingCart);
+        Mockito.verify(cartItemRepository, Mockito.times(1))
+                .save(cartItem);
+    }
+
+    @Test
+    @DisplayName("Check adding existing cart item to shopping cart "
+            + "and if shopping cart dto is returned")
+    void addCartItem_ExistingCartItem_ReturnsShoppingCartDto() {
+        //given
+        User user = new User();
+        user.setId(1L);
+        user.setEmail("email@gmail.com");
+        CartItem cartItem = new CartItem();
+        cartItem.setId(1L);
+        var requestDto = new CreateCartItemRequestDto(1L, 2);
+        cartItem.setQuantity(requestDto.quantity());
+
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setId(user.getId());
+        shoppingCart.setCartItems(Set.of(cartItem));
+        cartItem.setShoppingCart(shoppingCart);
+        CartItemDto cartItemDto = new CartItemDto(1L, "Test book", requestDto.quantity());
+
+        ShoppingCartDto expected = new ShoppingCartDto(shoppingCart.getId(),
+                shoppingCart.getId(),
+                Set.of(cartItemDto));
+
+        Mockito.when(bookRepository.existsById(requestDto.bookId())).thenReturn(true);
+        Mockito.when(cartItemRepository
+                        .findCartItemByShoppingCartIdAndBookId(user.getId(), requestDto.bookId()))
+                .thenReturn(Optional.of(cartItem));
+        Mockito.when(cartItemRepository.findByIdAndShoppingCartId(cartItem.getId(), user.getId()))
+                .thenReturn(Optional.of(cartItem));
+        Mockito.when(shoppingCartMapper.toShoppingCartDto(shoppingCart))
+                .thenReturn(expected);
+        Mockito.when(userRepository.findUserByEmail(user.getEmail()))
+                .thenReturn(Optional.of(user));
+        Mockito.when(authentication.getName()).thenReturn(user.getEmail());
+
+        //when
+        ShoppingCartDto actual = shoppingCartService.addCartItem(requestDto, authentication);
+
+        //then
+        Assertions.assertEquals(expected, actual);
+        Mockito.verify(cartItemRepository, Mockito.times(0))

```

Comment:
why 0?

Reply:
because if this item is already in the cart, then the quantity is updated, and the save method is not called